rootfs/etc/nginx/lua/balancer/leastconn.lua

```diff
@@ -0,0 +1,115 @@
+local setmetatable = setmetatable
+local string_format = string.format
+local ngx_log = ngx.log
+local INFO = ngx.INFO
+local WARN = ngx.WARN
+
+
+local _M = { name = "leastconn" }
+
+function _M.new(self, backend)
+    local o = {
+        peers = backend.endpoints
+    }
+    setmetatable(o, self)
+    self.__index = self
+    return o
+end
+
+function _M.is_affinitized()
+    return false
+end
+
+
+local function get_upstream_name(upstream)
+    return upstream.address .. ":" .. upstream.port
+end
+
+
+function _M.balance(self)
+    local peers = self.peers
+    local endpoint = peers[1]
+    local endpoints = ngx.shared.balancer_leastconn
+    local feasible_endpoints = {}
+
+    if #peers ~= 1 then
+        local lowestconns = 9999
+        -- find the lowest connection count
+        for _, peer in pairs(peers) do
+            local conns = endpoints:get(get_upstream_name(peer))
+            if conns == nil then
+              endpoints:set(get_upstream_name(peer),0,600)
+              conns = 0
+            end
+            ngx_log(WARN, "Found ", conns, " conns for peer ", get_upstream_name(peer))
+            if conns <= lowestconns then
+                lowestconns = conns
+            end
+        end
+
+        -- get peers with lowest connections
+        for _, peer in pairs(peers) do
+            local conns = endpoints:get(get_upstream_name(peer))
+            if conns ~= nil and conns == lowestconns then
+                feasible_endpoints[#feasible_endpoints+1] = peer
+            end
+        end
+        ngx_log(WARN, "got ", #feasible_endpoints, " feasible endpoints")
+
+        endpoint = feasible_endpoints[math.random(1,#feasible_endpoints)]
+    end
+
+
+    ngx_log(WARN, "chose endpoint ", get_upstream_name(endpoint))
+    -- Update the endpoint connection count with a TTL of 10 minutes
+    endpoints:incr(get_upstream_name(endpoint),1,1,600)
+
+    return get_upstream_name(endpoint)
+end
+
+function _M.after_balance(_)
+    local endpoints = ngx.shared.balancer_leastconn
+    local upstream = split.get_last_value(ngx.var.upstream_addr)
+
+    ngx_log(WARN, "decrement conn count for upstream ", upstream)
+
+    if util.is_blank(upstream) then
+        return
+    end
+    ngx_log(WARN, "decrement endpoints", upstream)
+    ngx_log(WARN, endpoints:incr(upstream,-1,0,600))
+end
+
+function _M.sync(self, backend)
+    local normalized_endpoints_added, normalized_endpoints_removed =
+        util.diff_endpoints(self.peers, backend.endpoints)
+
+    if #normalized_endpoints_added == 0 and #normalized_endpoints_removed == 0 then
+        ngx_log(WARN, "endpoints did not change for backend " .. tostring(backend.name))
+        return
+    end
+
+    ngx_log(WARN, string_format("[%s] peers have changed for backend %s", self.name, backend.name))
+
+    self.peers = backend.endpoints
+
+    for _, endpoint_string in ipairs(normalized_endpoints_removed) do
+        ngx.shared.balancer_leastconn:delete(endpoint_string)
```

Comment:
Removing these endpoints here may not have the desired affect. If a pod has a bunch of connections, briefly fails its readiness probe (causing it to no longer be included as a backend), we will remove it from being tracked here. But if the pod becomes ready again, it may still be handling a bunch of connections, but we will initialize its value to 0 again.

Maybe this is OK, but it is worth calling out.

Reply:
Yeah, that's not ideal behaviour. I think it would be safe to instead set expiration time to (say) 30 minutes. This would ensure we still track connection count for connections up to that time while not allowing the shared dict to grow indefinitely.

Reply:
Setting it to expire asynchronously makes some sense, though it will be difficult to choose the TTL.

It makes sense that this least conn load balancer would still track pods that are handling connections, even if the pod is not ready. It should decrement the number of connections on a backend that finishes a request, even if that pod is not ready for example.

Is there a way we can tell the difference between a pod that is not ready now (but therefore may become ready soon) versus a pod that is being killed (and therefore will never become ready again, and would be safe to fully delete)?

Reply:
> Is there a way we can tell the difference between a pod that is not ready now (but therefore may become ready soon) versus a pod that is being killed (and therefore will never become ready again, and would be safe to fully delete)?

Not to the best of my knowledge, and that's the heart of the issue - all we know is that an upstream address has disappeared from the list of endpoints for that service. We could find out more by querying the API server but that introduces substantially more complexity that I think the ingress-nginx team would not want to take on.

Reply:
This thread is still a bit concerning to me. It is unintuitive to me that a pod that becomes unready will result in ingress-nginx thinking it has 0 connections. Maybe setting the TTL to 2 days and just letting the entry in the map expire if we don't see it again in 2 days makes sense? (2 days is chosen somewhat arbitrarily)

Reply:
I've noticed some oddities here, while working with pods which had its readiness checks interrupted (due to it asking the same port/endpoint as the incoming requests), causing it to receive extra connections for a while

See the below graphs
<img width="1483" alt="image" src="https://github.com/kubernetes/ingress-nginx/assets/22740616/e20105b2-fc64-43ed-b716-ef62748de3b2">

Reply:
Some more correlation to this effect:
<img width="1486" alt="image" src="https://github.com/kubernetes/ingress-nginx/assets/22740616/b333c1e0-b656-4937-bed9-15a67d89936b">

Reply:
Update: all of these effects were due to removing nodes from the table when they "go missing", plus there was an off-by-one error that has now been fixed on my fork, see [the other thread](https://github.com/kubernetes/ingress-nginx/pull/9025#discussion_r1421357694)